pkg/validation/internal/bundle.go

```diff
@@ -133,23 +134,66 @@ func getOwnedCustomResourceDefintionKeys(csv *operatorsv1alpha1.ClusterServiceVe
 // - we could identify that the bundle size is close to the limit (bigger than 85%)
 func validateBundleSize(bundle *manifests.Bundle) []errors.Error {
 	warnPercent := 0.85
-	warnSize := int64(max_bundle_size * warnPercent)
+	warnSize := float64(max_bundle_size) * warnPercent
 	var errs []errors.Error
 
 	if bundle.CompressedSize == nil || *bundle.CompressedSize == 0 {
-		errs = append(errs, errors.WarnFailedValidation("unable to check the bundle size", nil))
+		errs = append(errs, errors.WarnFailedValidation("unable to check the bundle compressed size", bundle.Name))
 		return errs
 	}
 
+	if bundle.Size == nil || *bundle.Size == 0 {
+		errs = append(errs, errors.WarnFailedValidation("unable to check the bundle size", bundle.Name))
+		return errs
+	}
+
+	// From OPM (https://github.com/operator-framework/operator-registry) 1.17.5
+	// and OLM (https://github.com/operator-framework/operator-lifecycle-manager) : v0.19.0
+	// the total size checked is compressed
 	if *bundle.CompressedSize > max_bundle_size {
-		errs = append(errs, errors.ErrInvalidBundle(fmt.Sprintf("maximum bundle compressed size with gzip size exceeded: size=~%d MegaByte, max=%d MegaByte", *bundle.CompressedSize/(1<<(10*2)), max_bundle_size/(1<<(10*2))), nil))
-	} else if *bundle.CompressedSize > warnSize {
-		errs = append(errs, errors.WarnInvalidBundle(fmt.Sprintf("nearing maximum bundle compressed size with gzip: size=~%d MegaByte, max=%d MegaByte", *bundle.CompressedSize/(1<<(10*2)), max_bundle_size/(1<<(10*2))), nil))
+		errs = append(errs, errors.ErrInvalidBundle(
+			fmt.Sprintf("maximum bundle compressed size with gzip size exceeded: size=~%s , max=%s. Bundle uncompressed size is %s",
+				formatBytesInUnit(*bundle.CompressedSize),
+				formatBytesInUnit(max_bundle_size),
+				formatBytesInUnit(*bundle.Size)),
+			bundle.Name))
+	} else if float64(*bundle.CompressedSize) > warnSize {
+		errs = append(errs, errors.WarnInvalidBundle(
+			fmt.Sprintf("nearing maximum bundle compressed size with gzip: size=~%s , max=%s. Bundle uncompressed size is %s",
+				formatBytesInUnit(*bundle.CompressedSize),
+				formatBytesInUnit(max_bundle_size),
+				formatBytesInUnit(*bundle.Size)),
+			bundle.Name))
+	}
+
+	// Before these versions the bundles were not compressed
+	// and their size must be < ~1MB
+	if *bundle.Size > max_bundle_size {
+		errs = append(errs, errors.WarnInvalidBundle(
+			fmt.Sprintf("bundle uncompressed size exceeded the limit support for OLM versions relesed prior 1.17.5 :  size=~%s , max=%s. "+
+				"(these bundle cannot work in any cluster or vendor which uses OLM versions < 1.17.5 like OpenShift versions < 4.9)",
+				formatBytesInUnit(*bundle.Size),
+				formatBytesInUnit(max_bundle_size)),
+			bundle.Name))
```

Comment:
I am not sure of the general approach and the code above being in the API repository.
Here is how I see things.
- Operator authors should have the tooling to validate the size of the bundles when they create / publish them: "operator-sdk validate" and "opm index add". There messages like above are very valuable.
- Once the bundle gets loaded onto an OpenShift cluster there are two options:
  - OLM < 1.17.5 no compression and the bundle size should be less than 1MB. But I guess the code above will never run on such a cluster or is there the intent to port it back but not the uncompression mechanism? Either ways the resulting uncompressed size is to be checked and that's it. I am not sure what value the above message brings to a cluster administrator.
  - OLM >= 1.17.5 the message above is rather noise for a cluster administrator

Reply:
Hi @fgiloux, 

The checks added here does not end up for cluster admins. They are consumed by the tools that have the goal to build and validate bundles only. 

The reason to add this WARNING is to better dealing/avoid use cases like: https://bugzilla.redhat.com/show_bug.cgi?id=2054514

Some advantages with: 

- Authors be aware that their bundle can work only if it will be published/distributed in OLM > = 1.17.5 (OCP > 4.9). That shows very important info for who publishes in OperatorHub.io
-  CI ( e.g. CVP reports): Help us easily identify why the deployment tests fail. The warning will appear in the report and reduce our effort in troubleshooting these scenarios. 
- For operator authors: help identify and avoid scenarios like: https://bugzilla.redhat.com/show_bug.cgi?id=2054514 or any scenario where they are not aware that their bundle cannot work on this versions which will be still in use for a while. 
- Give the ability for pipelines to use these warn if they be unable to test the bundle against all versions where it will be deployed. For e.g. we have pipelines that are testing only in the upper OLM version where the bundle will be distributed because of the high cost to test against a big matrix. 
- Audit: Give the ability to verify how many use cases we have published in the index and are not working at all because the bundle is > 1MB and these indexes are used with OLM versions <  1.17.5 (OCP < 4.9)

Reply:
@camilamacedo86 I know the background and understand the reason for the check. There are surely useful for tools that build and validate bundles. The thing is that this code as part of the API repository is imported by operator-lifecycle-manager, registry, etc and I fail to see why.
This repository is diverging more and more from what is written on the tin.

Reply:
It. is important for many projects because it is a lib. 
The validator is not used outside of this these operations 
Also, re-design the API shows out of scope. The goal of this PR is only to add the required fix and improvements to the check.

Reply:
We spoke about this one and we agreed that we can close this one for now.Tks.